worker/caasapplicationprovisioner/application.go

```diff
@@ -97,20 +96,93 @@ func (a *appWorker) Wait() error {
 }
 
 func (a *appWorker) loop() error {
-	charmURL, err := a.facade.ApplicationCharmURL(a.name)
-	if errors.IsNotFound(err) {
-		a.logger.Debugf("application %q removed", a.name)
-		return nil
-	} else if err != nil {
-		return errors.Annotatef(err, "failed to get charm url for application")
-	}
-	charmInfo, err := a.facade.CharmInfo(charmURL.String())
+	// Wait till charm is upgraded to a v2 charm.
+	appStateWatcher, err := a.facade.WatchApplication(a.name)
 	if err != nil {
-		return errors.Annotatef(err, "failed to get application charm deployment metadata for %q", a.name)
+		return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
 	}
-	if charmInfo == nil ||
-		corecharm.Format(charmInfo.Charm()) < corecharm.FormatV2 {
-		return errors.Errorf("charm version 2 or greater required")
+	if err := a.catacomb.Add(appStateWatcher); err != nil {
+		return errors.Trace(err)
+	}
+	appStateChanges := appStateWatcher.Changes()
+	for {
+		format, err := a.charmFormat()
+		if errors.IsNotFound(err) {
+			a.logger.Debugf("application %q removed", a.name)
+			return nil
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+		if format >= corecharm.FormatV2 {
+			a.logger.Debugf("application %q is now a v2 charm", a.name)
+			break
+		}
+
+		appLife, err := a.facade.Life(a.name)
+		if errors.IsNotFound(err) {
+			a.logger.Debugf("application %q removed", a.name)
+			return nil
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+		if appLife == life.Dead {
+			a.logger.Debugf("application %q now dead", a.name)
+			return nil
+		}
+
+		// Wait for next app change, then loop to check charm format again.
+		select {
+		case <-appStateChanges:
+		case <-a.catacomb.Dying():
+			return a.catacomb.ErrDying()
+		}
+	}
+	appStateWatcher.Kill()
+	appStateChanges = nil
+
+	// If the application has an operator pod due to an upgrade-charm from a
+	// pod-spec charm to a sidecar charm, delete it. Also delete workload pod.
+	operatorDidExist := false
+	const maxDeleteLoops = 100
```

Comment:
Lets use "github.com/juju/retry" to power this loop. Currently we just spin as fast as possible.

Reply:
Using something like juju/retry is non-trivial, as this is not just a "call function with retries loop". It calls OperatorExists and only runs the deletes if it does exist, otherwise exits early. I tried forcing it into juju/retry but it would involve messy abuse of `IsFatalError`, I believe, unless I'm missing something. In any case, I've just added a `time.Sleep`.